cast: Ensure that pointers do not use rvalue / move overload

m-chaturvedi opened the pull request:
[Towards Issue 9398](https://github.com/RobotLocomotion/drake/issues/9398)

Requires:
- [x] #29 (CI fixes)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/pybind11/28)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
Thanks!
But it looks like the Travis CI bit needs updating; I'll submit a PR to bump to `master` to incorporate this:
https://github.com/pybind/pybind11/pull/1744

EricCousineau-TRI commented:
Submitted PR; updated this overview to indicate that it's a prerequisite.

EricCousineau-TRI reviewed (changes requested):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28)* status: all files reviewed, 1 unresolved discussion (waiting on @m-chaturvedi)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-LeSYTLhDBU9IxT3nJfh:-LeSYTLhDBU9IxT3nJfi:b3lzh0l) (no related file):*
In the downstream Drake PR, I'm still confused as to why we need `py_reference` when returning `T*` from a `cpp_function`.

Might you be able to add a unittest here in `pybind11` that tests returning / casting `T*` for a class with holder type `unique_ptr`, and a type that is non-copyable and non-moveable?

(Let me know if you'd like help on this part)

---


<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-LeSYTLhDBU9IxT3nJfh:-LeS_PES4WEV67t5j-6b:bu0nwi3) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

In the downstream Drake PR, I'm still confused as to why we need `py_reference` when returning `T*` from a `cpp_function`.

Might you be able to add a unittest here in `pybind11` that tests returning / casting `T*` for a class with holder type `unique_ptr`, and a type that is non-copyable and non-moveable?

(Let me know if you'd like help on this part)
</blockquote></details>

(And, uh, you may have already mentioned the reason beforehand... It's still fuzzy in my mind when reviewing the code / convos...)

---


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (changes requested):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28)* status: all files reviewed, 2 unresolved discussions (waiting on @m-chaturvedi)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-Lf1Q1hI7HneozzXNjWE:-Lf1Q1hI7HneozzXNjWF:b-l7xh9v) (no related file):*
Made some suggested changes: https://github.com/m-chaturvedi/pybind11/pull/1

Are you OK with these?

---


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28)* status: all files reviewed, 1 unresolved discussion (waiting on @m-chaturvedi)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-LeSYTLhDBU9IxT3nJfh:-Lf1Q6MK2q_daZX8f_Pc:b-a6ywps) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

(And, uh, you may have already mentioned the reason beforehand... It's still fuzzy in my mind when reviewing the code / convos...)
</blockquote></details>

OK I tried out an non-moveable and non-copyable type, same code paths. Thanks for your patience!

---


<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending remaining nits in meta-PR mentioned below

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-:-Lf1QmyKE6jVMIvaOFJN:b-rj5oyb)* status: all files reviewed, 1 unresolved discussion (waiting on @m-chaturvedi)



<!-- Sent from Reviewable.io -->

Commit:
cast: Ensure that pointers do not use rvalue / move overload

m-chaturvedi commented:
I apologize for not figuring this out before, but it looks like Pybind11 seems to provide some classes for testing.  We could forgo some of our code.  I tried to implement your comments in this:
https://github.com/m-chaturvedi/pybind11/commit/b893bd323e2192401c76cc57bb5efad0e310eee0
Let me know if this is acceptable.

EricCousineau-TRI reviewed (commented):
Those look good! Can you push them here?

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-:-Lf5oH2A9NRG0LFHIvBI:b-s2vdid)* status: all files reviewed, 1 unresolved discussion (waiting on @m-chaturvedi)



<!-- Sent from Reviewable.io -->

m-chaturvedi commented:
*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-Lf1Q1hI7HneozzXNjWE:-LfKmY2m53cShYqBx7wa:b-e1gkle) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Made some suggested changes: https://github.com/m-chaturvedi/pybind11/pull/1

Are you OK with these?
</blockquote></details>

Added the comments, the issue link and changed the message in the new commit.

---


<!-- Sent from Reviewable.io -->

m-chaturvedi reviewed (commented):
Done.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-:-LfKmUaz0bR7jv5Kik29:b-896fix)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @EricCousineau-TRI)



<!-- Sent from Reviewable.io -->

EricCousineau-TRI reviewed (approved):
Thanks! Sorry for the delay!

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/28#-:-Lf_4i0eEzjA2-vzYUsK:b-1isegc)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
